docs/architecture.md

```diff
@@ -0,0 +1,58 @@
+# Doctrine Extensions Architecture
+
+## Targeted Compatibility
+
+- PHP >= 7.2
+- Doctrine ORM 2.6+, 3.0 (no ODM / MongoDB - ORM only)
+- Doctrine Annotations Driver only
+
+## File Structure
+
+Doctrine Extensions consists of a single [monorepo](https://gomonorepo.org), that is
+automatically split into multiple read-only repositories for individual installation.
+
+Inside [`src`](../src) are three primary directories:
+
+1. [`Common`](../src/Common): common files required for any/all extensions to function.
+   Foundational files to support all extensions, hooks into Doctrine's functionality, etc.
+   
+2. [`Extension`](../src/Extension): the individual extensions, e.g. [Sluggable](../src/Extension/Sluggable)
+
+3. [`Tests`](../src/Tests): Common test utilities
+
+## Metadata
+
+At the source of any Doctrine project is
+[metadata](https://www.doctrine-project.org/projects/doctrine-orm/en/2.6/reference/metadata-drivers.html#metadata-drivers).
+
+Doctrine Extensions uses a custom Annotation driver to parse entities for their configuration.
+The resulting [metadata class](../src/Common/Metadata/ExtendedClassMetadata.php) holds
+the information needed for each extension to perform.
+
+> :rotating_light: Doctrine Extensions **only supports Annotations** at this time.
+> If you use another type of metadata driver (e.g. YAML) or you've configured a
+> different custom Annotation driver, this package is incompatible.
```

Comment:
Good point... we should look into decorating the annotation driver instead of directly extending it. That would provide better compatibility.

Reply:
Agreed! I think I'll upload something I created when I was learning how to create an extension myself. It uses the events system and totally separate classes.